File: packages/septr-go/tenant_aware.go
package septr

import (
	"strings"
)

func getNestedValue(obj map[string]interface{}, path string) interface{} {
	parts := strings.Split(path, ".")
	var current interface{} = obj
	for _, part := range parts {
		m, ok := current.(map[string]interface{})
		if !ok {
			return nil
		}
		current = m[part]
		if current == nil {
			return nil
		}
	}
	return current
}

func extractTenantFromJwt(claims map[string]string, jwtClaim string) string {
	if claims == nil {
		return ""
	}
	if val, ok := claims[jwtClaim]; ok {
		return val
	}
	return ""
}

type TenantLeak struct {
	Path  string      `json:"path"`
	Value interface{} `json:"value"`
}

func detectCrossTenantLeaks(expectedTenantID string, body interface{}, tenantColumn string) []TenantLeak {
	var leaks []TenantLeak

	var scan func(obj interface{}, path string)
	scan = func(obj interface{}, path string) {
		if obj == nil {
			return
		}

		switch v := obj.(type) {
		case map[string]interface{}:
			for key, val := range v {
				currentPath := path
				if currentPath == "" {
					currentPath = key
				} else {
					currentPath = currentPath + "." + key
				}

				if key == tenantColumn && val != nil {
					valStr := toString(val)
					if valStr != expectedTenantID {
						leaks = append(leaks, TenantLeak{Path: currentPath, Value: val})
					}
					continue
				}

				switch vt := val.(type) {
				case map[string]interface{}, []interface{}:
					scan(vt, currentPath)
				}
			}

		case []interface{}:
			for i, item := range v {
				scan(item, path+"["+toString(i)+"]")
			}
		}
	}

	scan(body, "")
	return leaks
}

func createTenantCheckResponse(tenantID string, body interface{}, config TenantAwareConfig) (bool, []TenantLeak) {
	leaks := detectCrossTenantLeaks(tenantID, body, config.TenantColumn)
	if len(leaks) > 0 && config.BlockOnMismatch {
		return true, leaks
	}
	return false, leaks
}

File: packages/septr-node/src/core/tenant-aware.ts
import type { TenantAwareConfig } from "./types"

function getNestedValue(obj: Record<string, unknown>, path: string): unknown {
  const parts = path.split(".")
  let current: unknown = obj
  for (const part of parts) {
    if (current === null || current === undefined || typeof current !== "object") return undefined
    current = (current as Record<string, unknown>)[part]
  }
  return current
}

export function extractTenantFromJwt(
  claims: Record<string, unknown> | null,
  jwtClaim: string,
): string | null {
  if (!claims) return null
  const value = getNestedValue(claims, jwtClaim)
  if (value === null || value === undefined) return null
  return String(value)
}

export interface TenantLeak {
  path: string
  value: unknown
}

export function detectCrossTenantLeaks(
  expectedTenantId: string,
  body: unknown,
  tenantColumn: string,
): TenantLeak[] {
  const leaks: TenantLeak[] = []

  function scan(obj: unknown, path: string): void {
    if (obj === null || obj === undefined) return
    if (typeof obj !== "object") return

    if (Array.isArray(obj)) {
      for (let i = 0; i < obj.length; i++) {
        scan(obj[i], `${path}[${i}]`)
      }
      return
    }

    const record = obj as Record<string, unknown>
    for (const key of Object.keys(record)) {
      const currentPath = path ? `${path}.${key}` : key
      const value = record[key]

      if (key === tenantColumn && value !== null && value !== undefined) {
        if (String(value) !== expectedTenantId) {
          leaks.push({ path: currentPath, value })
        }
        continue
      }

      if (typeof value === "object" && value !== null) {
        scan(value, currentPath)
      }
    }
  }

  scan(body, "")
  return leaks
}

export function createTenantCheckResponse(
  tenantId: string,
  body: unknown,
  config: TenantAwareConfig,
): { blocked: boolean; leaks: TenantLeak[] } {
  const leaks = detectCrossTenantLeaks(tenantId, body, config.tenantColumn)
  if (leaks.length > 0 && config.blockOnMismatch) {
    return { blocked: true, leaks }
  }
  return { blocked: false, leaks }
}
